tests/Avalonia.Controls.UnitTests/NameScopeTests.cs

```diff
@@ -20,36 +20,139 @@ public void Register_Registers_Element()
         }
 
         [Fact]
-        public void Unregister_Unregisters_Element()
+        public void Cannot_Register_New_Element_With_Existing_Name()
+        {
+            var target = new NameScope();
+
+            target.Register("foo", new object());
+            Assert.Throws<ArgumentException>(() => target.Register("foo", new object()));
+        }
+
+        [Fact]
+        public void Can_Register_Same_Element_More_Than_Once()
         {
             var target = new NameScope();
             var element = new object();
 
             target.Register("foo", element);
-            target.Unregister("foo");
+            target.Register("foo", element);
 
-            Assert.Null(target.Find("foo"));
+            Assert.Same(element, target.Find("foo"));
         }
 
         [Fact]
-        public void Cannot_Register_New_Element_With_Existing_Name()
+        public void Cannot_Register_New_Element_For_Completed_Scope()
         {
             var target = new NameScope();
+            var element = new object();
 
-            target.Register("foo", new object());
-            Assert.Throws<ArgumentException>(() => target.Register("foo", new object()));
+            target.Register("foo", element);
+            target.Complete();
+            Assert.Throws<InvalidOperationException>(() => target.Register("bar", element));
         }
 
+        object _found = null;
+        async void FindAsync(INameScope scope, string name)
```

Comment:
We should make our tests `async Task` instead of using `async void`.

Reply:
The continuation is expected to be run synchronously, so binding will receive its element right when it is registered. See https://github.com/AvaloniaUI/Avalonia/blob/b0111a2c488f42318e3e2bc2e6df9d6fc4217dfd/tests/Avalonia.Controls.UnitTests/NameScopeTests.cs#L94 which checks the value set by the continuation at particular stages of name scope interaction.

I guess I need to clarify that via a comment.

Reply:
Added comments